### PR TITLE
Token level timestamps for long-form generation in Whisper

### PR DESCRIPTION
**Pull request by zucchini-nlp**
# What does this PR do?

Continuation of PR #28984. Adds token level timestamps for long-form generation. The previous PR had a quite different of way to add timestamps, specifically by calling `extract_timestamps` for each segment and each batch separately. I believe, it can be done in one batch, and then divided into segments the same way sequences are divided.  

The final timestamps are already aligned with the total length, so there is not need to add start_time for each segment. Although, I am not sure if that is what we want to have, so I can remove this "total duration alignment" is needed.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [X] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [X] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@sanchit-gandhi 
@patrickvonplaten 
@gante ?

### CODE REVIEW

File: whisper.py [hunk 1]
```diff
@@ -0,0 +1,69 @@
+from transformers import WhisperForConditionalGeneration, WhisperProcessor
```
Comment (gante) [inline whisper.py:1]: This should not be part of the PR :D

**Review by gante — COMMENTED** [review]
In general it looks good to me, but I'd like to defer the approval to our Whisper expert @sanchit-gandhi :D

File: whisper.py (hunk 1 above)
Comment (zucchini-nlp) [reply whisper.py:1]: ooops

**Review by sanchit-gandhi — APPROVED** [review]
Thanks @zucchini-nlp - sounds good to me aligning the word-level timestamps on a per-segment level (rather than on the full sequence-level). We'll probably need to propagate some changes forward to the ASR pipeline so that it's compatible with these generation related changes, but we can do this in a follow-up PR

**Comment by sanchit-gandhi** [comment]
Also cc @ArthurZucker for Whisper-related timestamps

File: tests/models/whisper/test_modeling_whisper.py [hunk 2]
```diff
@@ -1969,6 +1969,56 @@ def test_tiny_token_timestamp_batch_generation(self):
...
+            [segment["token_timestamps"].shape for segment in segment_list]
+            for segment_list in generate_outputs["segments"]
+        ]
+        tokens_shape = [
+            [segment["tokens"].shape for segment in segment_list] for segment_list in generate_outputs["segments"]
+        ]
+        self.assertListEqual(tokens_shape, token_timestamps_shape)
+
+        # fmt: off
+        EXPECTED_OUTPUT = [
```
[the remark this answers is not in the document]
Comment (zucchini-nlp) [reply tests/models/whisper/test_modeling_whisper.py:2003]: Inspected from Transformers Whisper and deemed to be correct. I am not sure how the tests for short-form timestamps were written, so I just took an audio and copied model outputs.

File: tests/models/whisper/test_modeling_whisper.py [hunk 3]
```diff
@@ -1969,6 +1969,56 @@ def test_tiny_token_timestamp_batch_generation(self):
...
+            torch.tensor([44.7000, 44.7000, 44.8600, 44.9400, 45.1400, 45.1400, 45.2800, 45.6200, 45.9000, 46.2600, 47.1600, 47.4800, 47.7400, 48.1000, 48.2800, 48.4000, 48.6200, 48.8400, 49.0400, 49.2800, 49.4800, 49.6600, 49.9400, 50.5400]),
+            torch.tensor([50.5400, 50.5400, 50.6600, 50.8800, 51.2400, 51.7200, 52.8400]),
+            torch.tensor([52.9600, 52.9600, 53.0400, 53.2600, 53.4200, 53.5800, 53.9200, 54.1200, 54.7200, 54.9400, 55.2600, 55.6200, 55.9800, 56.5600, 56.8000, 56.9200, 57.3600, 57.9200, 58.1800, 58.5000, 58.6400, 58.8200]),
+            torch.tensor([58.6800, 58.6800, 59.1400, 59.5400, 59.9200, 60.1600, 60.3800, 60.8200, 61.6200, 62.2600, 75.2000]),
+        ]
+        # fmt: on
+
+        for segment, exp_segment in zip(generate_outputs["segments"][0], EXPECTED_OUTPUT):
+            self.assertTrue(torch.allclose(segment["token_timestamps"], exp_segment))
+
```
Comment (ArthurZucker) [inline tests/models/whisper/test_modeling_whisper.py:2021]: Could we add a small test like `test_return_timestamps_in_preprocess` in `/Users/arthurzucker/Work/transformers/tests/pipelines/test_pipelines_automatic_speech_recognition.py` just to make sure we have something explicit like 
```python 
                'chunks': [
                    {'text': ' Conquered', 'timestamp': (0.5, 1.2)},
                    {'text': ' returned', 'timestamp': (1.2, 1.64)},
                    {'text': ' to', 'timestamp': (1.64, 1.84)},
                    {'text': ' its', 'timestamp': (1.84, 2.02)},
                    {'text': ' place', 'timestamp': (2.02, 2.28)},
                    {'text': ' amidst', 'timestamp': (2.28, 2.8)},
                    {'text': ' the', 'timestamp': (2.8, 2.98)},
                    {'text': ' tents.', 'timestamp': (2.98, 3.48)},
                ],
```!

**Review by ArthurZucker — APPROVED** [review]
Let's add a small end to end test with explicit outputs, will help us in the long run! Otherwise LGTM

**Comment by zucchini-nlp** [comment]
@amyeroberts ready for review here
